examples/generative/midi_generation_with_transformer.py

```diff
@@ -0,0 +1,837 @@
+"""
+
+
+class MidiDataset(utils.PyDataset):
+    """A dataset for MIDI files that yields batches of input sequences and target sequences."""
+
+    def __init__(
+        self,
+        encoded_midis,
+        batch_size=CONFIG.batch_size,
+        max_sequence_length=CONFIG.max_sequence_length,
+        **kwargs,
+    ):
+        super(MidiDataset, self).__init__(**kwargs)
+        self.batch_size = batch_size
+        self.max_sequence_length = max_sequence_length
+        self.encoded_midis = encoded_midis
+        self._num_batches = (
+            len(encoded_midis) // batch_size + len(encoded_midis) % batch_size
+        )
+
+    def __len__(self):
+        """Get the number of batches."""
+        return self._num_batches
+
+    def __getitem__(self, idx):
+        """Generate random inputs and corresponding targets for the model."""
+        # Same as in the original paper, we always get a random batch.
+        # See: https://github.com/jason9693/MusicTransformer-tensorflow2.0/blob/f7c06c0cb2e9cdddcbf6db779cb39cd650282778/data.py
+        batch = random.sample(self.encoded_midis, k=self.batch_size)
+
+        # Convert the batch to sequences
+        batch_data = [
+            self._get_sequence(encoded_midi, self.max_sequence_length + 1)
+            for encoded_midi in batch
+        ]
+        batch_data = np.array(batch_data)
+
+        # Split the data into input and target sequences
+        x, y = batch_data[:, :-1], batch_data[:, 1:]
+        return x, y
+
+    def _get_sequence(self, data, max_length):
+        """Get a random sequence of notes from a file."""
+        # Truncate or pad the sequence
+        if len(data) > max_length:
+            start = random.randrange(0, len(data) - max_length)
+            data = data[start : start + max_length]
+        elif len(data) < max_length:
+            data = np.append(data, CONFIG.token_end_of_sentence)
+
+        # Pad the sequence if necessary
+        if len(data) < max_length:
+            data = np.concatenate(
+                (data, np.full(max_length - len(data), CONFIG.token_pad))
+            )
+        return np.asanyarray(data, dtype="int32")
+
+
+train_dataset, val_dataset = MidiDataset(train_midis), MidiDataset(val_midis)
+
+
+"""
+## Model definition
+
+It is time to define the model architecture. We use a Transformer decoder architecture 
+with a custom attention mechanism, relative global attention.
+
+### Relative Global Attention
+
+The following code implements the Relative Global Attention layer. It is used
+in place of the standard multi-head attention layer in the Transformer decoder.
+The main difference is that it includes a relative positional encoding that
+allows the model to learn relative positional information between tokens.
+"""
+
+
+@keras.utils.register_keras_serializable()
+class RelativeGlobalAttention(layers.Layer):
+    """
+    From Music Transformer (Huang et al., 2018)
+    [paper link](https://arxiv.org/pdf/1809.04281.pdf)
```

Comment:
In a code example class docstring, links are never rendered so no need for markdown here